internal/command/user_v2_human.go

```diff
@@ -0,0 +1,559 @@
+		if cmd != nil {
+			cmds = append(cmds, cmd)
+		}
+	} else {
+		// if phone not changed or empty, send no code
+		if phone.Number == "" || phone.Number == wm.Phone {
+			return cmds, code, err
+		}
+		c, err := c.newPhoneCode(ctx, c.eventstore.Filter, alg) //nolint:staticcheck
+		if err != nil {
+			return cmds, code, err
+		}
+		cmds = append(cmds, user.NewHumanPhoneCodeAddedEventV2(ctx, &wm.Aggregate().Aggregate, c.Crypted, c.Expiry, phone.ReturnCode))
+		if phone.ReturnCode {
+			code = &c.Plain
+		}
+	}
+	return cmds, code, nil
+}
+
+func changeUserProfile(ctx context.Context, cmds []eventstore.Command, wm *UserV2WriteModel, profile *Profile) ([]eventstore.Command, error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.End() }()
+
+	cmd, err := wm.NewProfileChangedEvent(ctx, profile.FirstName, profile.LastName, profile.NickName, profile.DisplayName, profile.PreferredLanguage, profile.Gender)
+	if cmd != nil {
+		return append(cmds, cmd), err
+	}
+	return cmds, err
+}
+
+func (c *Commands) changeUserPassword(ctx context.Context, cmds []eventstore.Command, wm *UserV2WriteModel, password *Password, alg crypto.EncryptionAlgorithm) ([]eventstore.Command, error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.End() }()
+
+	// Either have a code to set the password
+	if password.PasswordCode != nil {
+		if err := crypto.VerifyCodeWithAlgorithm(wm.PasswordCodeCreationDate, wm.PasswordCodeExpiry, wm.PasswordCode, *password.PasswordCode, alg); err != nil {
+			return cmds, err
+		}
+	}
+	var encodedPassword string
+	// or have the old password to change it
+	if password.OldPassword != nil {
+		// newly encode old password if no new and already encoded password is set
+		pw := *password.OldPassword
+		if password.Password != nil {
+			pw = *password.Password
+		}
+		alreadyEncodedPassword, err := c.verifyAndUpdatePassword(ctx, wm.PasswordEncodedHash, *password.OldPassword, pw)
+		if err != nil {
+			return cmds, err
+		}
+		encodedPassword = alreadyEncodedPassword
+	}
+
+	// password already hashed in request
+	if password.EncodedPasswordHash != nil {
+		cmd, err := c.setPasswordCommand(ctx, &wm.Aggregate().Aggregate, wm.UserState, *password.EncodedPasswordHash, password.ChangeRequired, true)
+		if cmd != nil {
+			return append(cmds, cmd), err
+		}
+		return cmds, err
+	}
+	// password already hashed in verify
+	if encodedPassword != "" {
+		cmd, err := c.setPasswordCommand(ctx, &wm.Aggregate().Aggregate, wm.UserState, encodedPassword, password.ChangeRequired, true)
+		if cmd != nil {
+			return append(cmds, cmd), err
+		}
+		return cmds, err
+	}
+	// password still to be hashed
+	if password.Password != nil {
+		cmd, err := c.setPasswordCommand(ctx, &wm.Aggregate().Aggregate, wm.UserState, *password.Password, password.ChangeRequired, false)
+		if cmd != nil {
+			return append(cmds, cmd), err
+		}
+		return cmds, err
+	}
+	// no password changes necessary
+	return cmds, nil
```

Comment:
same as with the changePhone/Email functions... it be easier be easier to split the function into handling the change with an old password and one handling the code
currently the logic is spread in the whole function, e.g. if an old password is provided for verification, that will already encode the new one if it is not passed encoded, but the event is only created afterwards. Not to mention that, if the new password is already encoded, to check the old password, the old password will also be passed to `verifyAndUpdatePassword` just to satisfy that function and discard the newly hashed password as it is not intended to be used

Reply:
The problem here is that it doesn't matter if I provide a new password or not, some password will be hashed anyway, which is a problem with the password hasher, which is inside verifyAndUpdatePassword.
- old password and new password already encoded -> old password gets hashed again if no new is provided

Tried to at least structure it for the encoded passwords, but otherwise I would also have to change the password hasher. (Verify also uses VerifyAndUpdate just with the password twice)